File: cs50 2019/pset2/vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <stdlib.h>

int shift(char c);

int main(int argc, string argv[])
{
    // cheking the input
    if (argc != 2) {
        printf("Usage: ./vigenere key\n");
        return 1;
    }
    for (int i = 0; i<strlen(argv[1]); i++) {
        if (isdigit(argv[1][i])) {
            printf("Usage: ./vigenere key\n");
            return 1;
        }
    }

    string plaintext = get_string("plaintext: ");
    printf("ciphertext: ");
    int key;
    for (int i = 0, j = 0; i<strlen(plaintext); i++, j++) {
        if (isupper(plaintext[i]) || islower(plaintext[i])) {
            key = shift(argv[1][j % strlen(argv[1])]);
        } else {
            j--;
        }
        if (isupper(plaintext[i])) {
            printf("%c", ((plaintext[i]-65 + key) % 26) + 65);
        } else if (islower(plaintext[i])) {
            printf("%c", ((plaintext[i]-97 + key) % 26) + 97);
        } else {
            printf("%c", plaintext[i]);
        }
    }
    printf("\n");
}


int shift(char c) {
   if (isupper(c)) {
       return c - 65;
   } else if (islower(c)) {
       return c - 97;
   } else {
       return 0;
   }
}

File: cs50 2019/pset6/vigenere.py
from sys import argv, exit
from cs50 import get_string


def main():
    key_word = argv[1]
    if len(argv) != 2:
        exit(1)
    for i in key_word:
        if not i.isalpha():
            exit(1)

    plaintext = get_string("plaintext: ")
    print("ciphertext: ", end='')

    i = j = key = 0
    while i < len(plaintext):
        value = plaintext[i]

        if value.isalpha():
            key = determine_shift(key_word[j % len(key_word)])
        else:
            j -= 1

        if value.isupper():
            print(chr(((ord(value) - 65 + key) % 26) + 65), end='')
        elif value.islower():
            print(chr(((ord(value) - 97 + key) % 26) + 97), end='')
        else:
            print(value, end='')

        i += 1
        j += 1
    print()


def determine_shift(char):
    if char.isupper():
        return ord(char) - 65
    elif char.islower():
        return ord(char) - 97
    return 0


if __name__ == "__main__":
    main()
